Index: framework/entity/src/org/ofbiz/entity/connection/DBCPConnectionFactory.java
===================================================================
--- framework/entity/src/org/ofbiz/entity/connection/DBCPConnectionFactory.java	(revision 1447126)
+++ framework/entity/src/org/ofbiz/entity/connection/DBCPConnectionFactory.java	(working copy)
@@ -51,6 +51,8 @@
 
     public static final String module = DBCPConnectionFactory.class.getName();
     protected static Map<String, ManagedDataSource> dsCache = FastMap.newInstance();
+    protected static Map<String, XAConnectionFactory> xacfCache = FastMap.newInstance();
+    protected static Map<String, GenericObjectPool> gopCache = FastMap.newInstance();
 
     public Connection getConnection(GenericHelperInfo helperInfo, Element jdbcElement) throws SQLException, GenericEntityException {
         ManagedDataSource mds = dsCache.get(helperInfo.getHelperFullName());
@@ -174,15 +176,29 @@
 
             // cache the pool
             dsCache.put(helperInfo.getHelperFullName(), mds);
+            xacfCache.put(helperInfo.getHelperFullName(), xacf);
+            gopCache.put(helperInfo.getHelperFullName(), pool);
 
             return TransactionFactory.getCursorConnection(helperInfo, mds.getConnection());
         }
     }
+    
+    public void removeConnection(GenericHelperInfo helperInfo) {
+        dsCache.remove(helperInfo.getHelperFullName());
+    }
 
     public void closeAll() {
         // no methods on the pool to shutdown; so just clearing for GC
         dsCache.clear();
     }
+    
+    public XAConnectionFactory getXAConnectionFactory(GenericHelperInfo helperInfo) {
+        return xacfCache.get(helperInfo.getHelperFullName());
+    }
+    
+    public GenericObjectPool getGenericObjectPool(GenericHelperInfo helperInfo) {
+        return gopCache.get(helperInfo.getHelperFullName());
+    }
 
     public static Map getDataSourceInfo(String helperName) {
         Map dataSourceInfo = new HashMap();
Index: framework/entity/src/org/ofbiz/entity/jdbc/DatabaseUtil.java
===================================================================
--- framework/entity/src/org/ofbiz/entity/jdbc/DatabaseUtil.java	(revision 1447126)
+++ framework/entity/src/org/ofbiz/entity/jdbc/DatabaseUtil.java	(working copy)
@@ -83,6 +83,7 @@
 
     boolean isLegacy = false;
     protected ExecutorService executor;
+    protected List<Connection> connections = FastList.newInstance();
 
     // OFBiz DatabaseUtil
     public DatabaseUtil(GenericHelperInfo helperInfo) {
@@ -145,6 +146,9 @@
                 throw new GenericEntityException("No connection avaialble for URL [" + connectionUrl + "]");
             }
         }
+        
+        connections.add(connection);
+        
         if (!TransactionUtil.isTransactionInPlace()) {
             connection.setAutoCommit(true);
         }
@@ -3203,6 +3207,18 @@
             }
         }
     }
+    
+    public void close() {
+        for (Connection connection : connections) {
+            try {
+                if (!connection.isClosed()) {
+                    connection.close();
+                }
+            } catch (SQLException e) {
+                Debug.logWarning("Could not close connection: " + connection, module);
+            }
+        }
+    }
 
     /* ====================================================================== */
     /* ====================================================================== */
Index: framework/entityext/src/org/ofbiz/entityext/data/EntityDataServices.java
===================================================================
--- framework/entityext/src/org/ofbiz/entityext/data/EntityDataServices.java	(revision 1447126)
+++ framework/entityext/src/org/ofbiz/entityext/data/EntityDataServices.java	(working copy)
@@ -407,6 +407,8 @@
         // step 8 - checkdb
         Debug.logImportant("Running DB check with add missing enabled", module);
         dbUtil.checkDb(modelEntities, messages, true);
+        
+        dbUtil.close();
 
         Map<String, Object> result = ServiceUtil.returnSuccess();
         result.put("messages", messages);
