Index: framework/common/webcommon/login.ftl
===================================================================
--- framework/common/webcommon/login.ftl	(revision 1425813)
+++ framework/common/webcommon/login.ftl	(working copy)
@@ -42,6 +42,7 @@
             <td class="label">${uiLabelMap.CommonPassword}</td>
             <td><input type="password" name="PASSWORD" value="" size="20"/></td>
           </tr>
+          <#--
           <#if ("Y" == useMultitenant) >
               <#if !requestAttributes.tenantId?exists>
                   <tr>
@@ -52,6 +53,7 @@
                   <input type="hidden" name="tenantId" value="${requestAttributes.tenantId?if_exists}"/>
               </#if>
           </#if>
+          -->
           <tr>
             <td colspan="2" align="center">
               <input type="submit" value="${uiLabelMap.CommonLogin}"/>
